### PR TITLE
Issue #2195 - Enable client checkpoint driven reindex

### PR DESCRIPTION
**Pull request by tbieste**
Issue #2195 - Enable client checkpoint driven reindex

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: docs/src/pages/guides/FHIRPerformanceGuide.md
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRPerformanceGuide.md:203]: seems a bit redundant now :-)

Comment (lmsurpre) [reply docs/src/pages/guides/FHIRPerformanceGuide.md:203]:
```suggestion
| $reindex           | One HTTP call can request multiple resources to be reindexed. By default, each resource is reindexed in the scope of its own transaction. Reindexing is a relatively quick operation per resource - usually well under 1s - so transaction timeouts are unlikely. However, if a list of logical resource IDs is specified, all those resources will be reindexed within a single tranaction, so reduce the number logical resource IDs specified if transaction timeouts occur. Use concurrent requests to increase overall throughput. |
```

Comment (tbieste) [reply docs/src/pages/guides/FHIRPerformanceGuide.md]: Updated.

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRSearchConfiguration.md:281]:
```suggestion
Another option is to have the client utilize the `$retrieve-index` and `$reindex` in parallel to drive the processing.
```

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRSearchConfiguration.md:300]:
```suggestion
To aid in the re-indexing process, the IBM FHIR Server team has expanded the fhir-bucket resource-loading tool to support driving the reindex, with the option of using either the server-side-driven or client-side-driven approach. The fhir-bucket tool uses a thread-pool to make concurrent POST requests to the IBM FHIR Server `$retrieve-index` and `$reindex` custom operations.
```

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/app/Main.java
Comment (lmsurpre) [inline fhir-bucket/src/main/java/com/ibm/fhir/bucket/app/Main.java:860]:
```suggestion
        }
```

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRSearchConfiguration.md:295]: is this our database-internal ids or the resource ids?
if the former, we need to change the name... "logical resource id" has a specific meaning to FHIR users and this aint it.

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (lmsurpre) [inline docs/src/pages/guides/FHIRSearchConfiguration.md:276]: will need to get reconciled with https://github.com/IBM/FHIR/pull/2489

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java
Comment (lmsurpre) [inline fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:74]: Does it use `the number of cpu cores on the system - 1` in the pool?
Doe it relate to the passed `--reindex-concurrent-requests` value at all?

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (lmsurpre) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:295]: I do realize that this is what its called in our database (as documented at https://github.com/IBM/FHIR/blob/main/fhir-persistence-schema/docs/SchemaMigration.md#logical_resources-and-resources) but I really don't want to expose this to our users because it is really confusing.  Happy to brainstorm alternative names as needed.

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (tbieste) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:281]: Fixed.

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (tbieste) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:300]: Fixed.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/app/Main.java
Comment (tbieste) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/app/Main.java:860]: Fixed.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java
Comment (tbieste) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:74]: Yes, the passed `--reindex-concurrent-requests` value determines the number of worker threads created in the thread pool in the monitorLoop method.  This was same logic as used in the ServerDrivenReindexOperation, so I followed that.

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (tbieste) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:295]: I could just change the name to something like "indexId", so users just see it as just some ID used in reindex processing.

Comment (tbieste) [reply docs/src/pages/guides/FHIRSearchConfiguration.md]: Updated to use the name "indexIds" for the parameter, instead of "logicalResourceIds".  This should avoid confusing user into thinking it is the same as resource IDs.

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (tbieste) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:276]: Yeah, those merge conflicts will need to be resolved.  Since the upstream branch `robin-perf-eval` is pretty far behind main right now; I wonder if that may be easy to do once `robin-perf-eval` is brought up-to-date with those commits.

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (lmsurpre) [reply docs/src/pages/guides/FHIRSearchConfiguration.md]: I am happy with `indexIds`...this is WAY better than `logicalResourceIds` :-)

**Review by lmsurpre — APPROVED** [review]
LGTM.  The test failures are expected until #2490 is merged.

File: operation/fhir-operation-reindex/src/main/resources/retrieve-index.json
Comment (lmsurpre) [inline operation/fhir-operation-reindex/src/main/resources/retrieve-index.json:17]: If we support this one at the type level, that would allow us to add selective reindexing to the fhir-bucket client:  https://github.com/IBM/FHIR/issues/2525

In #2489 we did the same thing for the `$reindex` operation (i.e. made it work on the type level in addition to the system level).

Comment (tbieste) [reply operation/fhir-operation-reindex/src/main/resources/retrieve-index.json:17]: Yeah, that makes sense.  I can add `$retrieve-index` support at the type level.

Comment (tbieste) [reply operation/fhir-operation-reindex/src/main/resources/retrieve-index.json]: Added.

File: operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java
Comment (lmsurpre) [inline operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:154]: is it safe to remove this now that we're checking it in AbstractOperation?

Comment (lmsurpre) [reply operation/fhir-operation-reindex/src/main/java/com/ibm/fhir/operation/reindex/ReindexOperation.java:154]: never mind, forgot that we still allow users to pass it in via that logicalResourceId param

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (punktilious) [inline docs/src/pages/guides/FHIRSearchConfiguration.md:283]: repeatedly

Comment (tbieste) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:283]: Fixed.

File: fhir-bucket/README.md
Comment (punktilious) [inline fhir-bucket/README.md:362]: repeatedly is repeated

File: fhir-bucket/README.md
Comment (punktilious) [inline fhir-bucket/README.md:366]: Do we want to mention when to stop client-driven reindexing? This is only relevant if ingestion is started again after an upgrade/config update but before reindex has completed. The newly ingested resources won't need to be reindexed, so we could stop earlier. Perhaps there should be an option for an "end-stop" timestamp with fhir-bucket. It's perhaps not critical, unless the new ingestion workload is high and you are trying to get reindex to complete as quickly as possible.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java
Comment (punktilious) [inline fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:252]: Not sure this return comment lines up with the logic in the method. Not sure I understand how queued and active are working together here. I'd like to walk through this. I'd like to understand what happens when ids are rejected by the blockingQueue.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java
Comment (punktilious) [inline fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:313]: This could limit throughput if threads are frequently sleeping waiting for new work. Better to use locks and conditions.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/ReindexResourceDAO.java
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/ReindexResourceDAO.java:51]: The is_deleted flag is now available on the LOGICAL_RESOURCES table.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/RetrieveIndexDAO.java
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/RetrieveIndexDAO.java:63]: To avoid any odd behavior by the optimizer, we should use the cached resourceTypeId value and give it as a literal instead of joining against the resource_types table. It will be slightly quicker.

**Review by punktilious — APPROVED** [review]
LGTM if we either address the performance comments or create issues

File: fhir-bucket/README.md
Comment (tbieste) [reply fhir-bucket/README.md:362]: Fixed.

File: fhir-bucket/README.md
Comment (tbieste) [reply fhir-bucket/README.md:366]: Since the "notModifiedAfter" parameter is set to the timestamp the reindex is started, any newly ingested resource won't get reindex; so that should already be handled.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java
Comment (tbieste) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:252]: If the IDs are rejected by the blockingQueue after waiting up to 10 seconds, then "queued = false", so `lastIndexId` is not updated.  So the next time retrieveIndex is called it calls right where it last left off.  I can change OFFER_TIMEOUT_IN_SEC to a bigger number (i.e. 30 seconds), I just didn't want it too high so if signalStop() is called while the queue is full, it doesn't take so long to shut everything it down.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java
Comment (tbieste) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:313]: Replaced this with a call to blockingQueue.poll(...), so it doesn't use this arbitrary sleep to wait to check if the queue is still empty.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/ReindexResourceDAO.java
Comment (tbieste) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/ReindexResourceDAO.java:51]: Updated.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/RetrieveIndexDAO.java
Comment (tbieste) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/RetrieveIndexDAO.java:67]: Updated.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java
Comment (tbieste) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java]: The return true, is just for if "active" is false (i.e. the worker threads are not active and consuming work), we don't want to just keep trying to enqueue more work, but rather return back to the monitorLoop() to start up the worker threads (or shutdown if running is false).

File: docs/src/pages/guides/FHIRSearchConfiguration.md
Comment (tbieste) [reply docs/src/pages/guides/FHIRSearchConfiguration.md:276]: Changed merged.

**Comment by tbieste** [comment]
I think I have all the comments addressed.  Ready to merge when we have a clean build.

**Pull request merged**
